wellcomeml/ml/bert_semantic_equivalence.py

```diff
@@ -198,15 +221,191 @@ def predict(self, X):
         return self.score(X).argmax(axis=1)
 
     def save(self, path):
+        """Saves model to path"""
         os.makedirs(path, exist_ok=True)
         self.model.save_pretrained(path)
 
     def load(self, path):
+        """Loads model from path"""
         self._initialise_models()
         self.model = TFBertForSequenceClassification.from_pretrained(path)
         self.trained_ = True
 
 
+class SemanticMetaBert(SemanticEquivalenceClassifier):
+    """
+    Class to fine tune Semantic Classifier, allowing the possibility of
+    adding metadata to the classification. Extends
+    SemanticEquivalenceClassifier, and extends modules for data prep
+    and model initialisation. Fit, predict, score remain intact
+    """
+    def __init__(self, n_numerical_features, dropout_rate=0.1, **kwargs):
+        """
+        Initialises SemanticMetaBert, with n_numerical_features
+
+        Args:
+            n_numerical_features(int): Number of features of the model which
+            are not text
+            dropout_rate(float): Dropout rate after concatenating features
+            (default = 0.1)
+            **kwargs: Any kwarg to `SemanticEquivalenceClassifier`
+        """
+        super().__init__(**kwargs)
+        self.n_numerical_features = n_numerical_features
+        self.dropout_rate = dropout_rate
+        self.meta_model = None
+
+    def _separate_features(self, X):
+        X_text = [x[:self.n_numerical_features] for x in X]
+        X_numerical = [x[self.n_numerical_features:] for x in X]
+
+        return X_text, X_numerical
+
+    def _initialise_models(self):
+        """Extends/overrides super class initialisation of model"""
+        super()._initialise_models()
+
+        # Define text input features
+        text_features = {"input_ids", "attention_mask", "token_type_ids"}
+        input_text_tensors = [
+            tf.keras.layers.Input(
+                name=feature_name,
+                shape=tf.TensorShape([None]),
+                dtype=tf.int32
+            ) for feature_name in text_features
+        ]
+
+        input_numerical_data_tensor = tf.keras.layers.Input(
+            name="numerical_metadata",
+            shape=tf.TensorShape([self.n_numerical_features]),
+            dtype=tf.float32
+        )
+        # Calls the CLS layer of Bert
+        x = self.model.layers[0](input_text_tensors)[1]
+
+        # Concatenates with numerical features
+        x = tf.keras.layers.concatenate([x, input_numerical_data_tensor],
+                                        name='concatenate')
+
+        # Drop out layer to the concatenated features
+        x = tf.keras.layers.Dropout(rate=self.dropout_rate,
+                                    name='concatenated_dropout')(x)
+
+        # Dense layer that will be used for softmax prediction later
+        x = tf.keras.layers.Dense(2, name='dense_layer')(x)
+
+        self.meta_model = tf.keras.Model(
+            input_text_tensors + [input_numerical_data_tensor],
```

Comment:
why the need for the parenthesis in the second? maybe it would be better for those variables to have the same format from the beggining

Reply:
+1: this was now changed